python/llm/src/bigdl/llm/transformers/models/utils.py

```diff
@@ -46,3 +48,21 @@ def append_kv_cache(cache_k, cache_v, key_states, value_states):
     new_cache_v = cache_v.as_strided(new_size, cache_v.stride(), storage_offset=0)
     new_cache_v[:, :, cache_v.size(2):cache_k.size(2) + key_states.size(2), :] = value_states
     return new_cache_k, new_cache_v
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., :x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
```

Comment:
GPT-J uses a slightly different rotation method which rotates every two consecutive elements instead of rotating first half and second half like this one. (most models rotate half) https://github.com/huggingface/transformers/blob/v4.33.2/src/transformers/models/gptj/modeling_gptj.py#L69

Should we add a parameter in `apply_rotary_pos_emb` or modify the name `apply_rotary_pos_emb` so that others will take a closer look at the function before applying it.

Reply:
Thanks for reminder! Have added a `model_family` parameter to `apply_rotary_pos_emb` .